File: tests/Spreads.Tests/Collections/SortedMapTests.fs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.


namespace Spreads.Tests.Collections

open FsUnit
open NUnit.Framework

open System
open System.Linq
open System.Collections.Generic
open System.Diagnostics
open Spreads
open Spreads.Collections
open System.Threading
open System.Threading.Tasks


[<TestFixture>]
type SortedMapTests() =

    [<Test>]
    member this.``Could create SM of irregular keys and values``() =
        let keys = [|1;2;3;5;6;8;9;15|]
        let values =  [|1;2;3;5;6;8;9;15|]
        let sm = SortedMap.OfSortedKeysAndValues(keys, values)
        Assert.IsTrue(sm.Keys.SequenceEqual(keys))
        Assert.IsTrue(sm.Values.SequenceEqual(values))
        Assert.IsTrue(not sm.IsRegular)
        Assert.IsTrue(sm.keys.Length = keys.Length)
        ()

    [<Test>]
    member this.``Could create SM of regular keys and values``() =
        let keys = [|1;2;3;4;5;6;7;8|]
        let values =  [|1;2;3;4;5;6;7;8|]
        let sm = SortedMap.OfSortedKeysAndValues(keys, values)
        Assert.IsTrue(sm.Keys.SequenceEqual(keys))
        Assert.IsTrue(sm.Values.SequenceEqual(values))
        Assert.IsTrue(sm.IsRegular)
        Assert.IsTrue(sm.keys.Length = 2)
        ()

    [<Test>]
    member this.``Sorted list inserts``() =
        let watch = Stopwatch.StartNew();
        let sm = SortedList<DateTime, float>()
        let lim = 1000000L
        let td = DateTime.Today
        for i in 0L..lim do
            let dt = td.AddTicks(i)
            sm.[dt] <- float(i)
        
        watch.Stop()
        Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())

        ()



    [<Test>]
    member this.``Sorted map inserts``() =
        for c in 0..9 do
            let mutable watch = Stopwatch.StartNew();
            let sm = SortedMap<int64, float>()
            let lim = 1000000L
            for i in 0L..lim do
                sm.[i] <- float(i)
        
            watch.Stop()
            Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            watch <- Stopwatch.StartNew();

            let mutable sum = 0.0
            for i in 0L..lim do
                sum <- sm.[i]
        
            watch.Stop()
            Console.WriteLine(">Reads per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            watch <- Stopwatch.StartNew();
            let ptr = sm.GetCursor()
            let mutable sum = 0.0
            for i in 0L..lim do
                ptr.MoveNext() |> ignore
                sum <- ptr.CurrentValue
        
            watch.Stop()
            Console.WriteLine(">Iterations per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            ()


    [<Test>]
    member this.``Sorted map indexed reads``() =
        for c in 0..9 do
            let mutable watch = Stopwatch.StartNew();
            let sm = SortedMap<int64, float>()
            let lim = 1000000L
            for i in 0L..lim do
                sm.[i] <- float(i)
        
            watch.Stop()
            Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            watch <- Stopwatch.StartNew();

            let mutable sum = 0.0
            for i in 0L..lim do
                sum <- sm.values.[int(i)]
        
            watch.Stop()
            Console.WriteLine(">Indexed reads per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            watch <- Stopwatch.StartNew();

            let mutable sum = 0.0
            for i in 0L..lim do
                sm.values.[int(i)] <- float(i)
        
            watch.Stop()
            Console.WriteLine(">Indexed sets per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            ()

    [<Test>]
    member this.``Dictionary inserts and reads``() =
        for c in 0..9 do
            let mutable watch = Stopwatch.StartNew();
            let sm = Dictionary<int64, float>()
            let lim = 1000000L
            for i in 0L..lim do
                sm.[i] <- float(i)
        
            watch.Stop()
            Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())


            watch <- Stopwatch.StartNew();

            let mutable sum = 0.0
            for i in 0L..lim do
                sum <- sm.[i]
        
            watch.Stop()
            Console.WriteLine(">Reads per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())

            ()

    [<Test>]
    member this.``Could remove data from SM``() =
        let mutable watch = Stopwatch.StartNew();
        let sm = SortedMap<int64, float>()
        let lim = 100000L
        for i in 0L..lim do
          if i <> 2L then sm.[i] <- float(i) // irregular
        sm.TrimExcess()
        let okf, fs = sm.RemoveFirst()

        let okl, ls = sm.RemoveLast()

        ()

//    [<Test>]
//    member this.``IImmutableOrderedMap reverse``() =
//
//        let mutable watch = Stopwatch.StartNew();
//        let hsm:IImmutableOrderedMap<int,int> = SortedMap<int, int>() :> IImmutableOrderedMap<int,int>
//        let lim = 5120
//        for i = 0 to lim  do
//            (hsm :?> IOrderedMap<int,int>).Add(i,i)
//            //sm.[i] <- sm.Add()
//        
//        watch.Stop()
//        Console.WriteLine(">Mutable inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())
//
//        watch <- Stopwatch.StartNew();
//        let mutable hsm:IImmutableOrderedMap<int,int> = SortedMap<int, int>() :> IImmutableOrderedMap<int,int>
//        let lim = 5120
//        for i = 0 to lim  do
//            hsm <- hsm.Add(i,i)
//        
//        watch.Stop()
//        Console.WriteLine(">Immutable inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())
//
//        ()

    [<Test>]
    member this.``Nested sorted list``() =
        for c in 0..9 do
            let watch = Stopwatch.StartNew();
            let nsl = SortedList<int64, SortedList<uint16, float>>()
            let lim = 1000000L
            let bucket_size = 512
            for i = int(0L) to  int(lim) do
                let hash = int64(i / bucket_size)
                let idx = i % bucket_size
                let tg = nsl.TryGetValue(hash)
                if fst tg then 
                    (snd tg).[uint16(idx)] <- float(i)
                else 
                    let newSm = SortedList()
                    newSm.[uint16(idx)] <- float(i)
                    nsl.[hash] <- newSm
        
            watch.Stop()
            Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())

            ()

    [<Test>]
    member this.``Nested sorted map``() =
        for c in 0..9 do
            let watch = Stopwatch.StartNew();
            let nsl = SortedMap<int64, SortedMap<uint16, float>>()
            let lim = 1000000L
            let bucket_size = 512
            for i = int(0L) to  int(lim) do
                let hash = int64(i / bucket_size)
                let idx = i % bucket_size
                let tg = nsl.TryGetValue(hash)
                if fst tg then 
                    (snd tg).[uint16(idx)] <- float(i)
                else 
                    let newSm = SortedMap()
                    newSm.[uint16(idx)] <- float(i)
                    nsl.[hash] <- newSm
        
            watch.Stop()
            Console.WriteLine(">Inserts per sec: " + (1000L * int64(lim)/watch.ElapsedMilliseconds).ToString())

            ()

File: tests/Spreads.Collections.Tests/SortedMapTests.cs
//// This Source Code Form is subject to the terms of the Mozilla Public
//// License, v. 2.0. If a copy of the MPL was not distributed with this
//// file, You can obtain one at http://mozilla.org/MPL/2.0/.


//using System;
//using System.Collections.Generic;
//using System.Diagnostics;
//using System.Threading.Tasks;
//using NUnit.Framework;
//using Spreads.Serialization;

//namespace Spreads.Collections.Tests {

//    [TestFixture]
//    public class SortedMapTests {

//        [SetUp]
//        public void Init() {
//        }

//        [Test]
//        public void CouldEnumerateGrowingSM() {
//            var count = 1000000;
//            var sw = new Stopwatch();
//            sw.Start();
//            var sm = new SortedMap<DateTime, double>();
//            //sm.IsSynchronized = true;
//            var c = sm.GetCursor();

//            for (int i = 0; i < count; i++) {
//                sm.Add(DateTime.UtcNow.Date.AddSeconds(i), i);
//                c.MoveNext();
//                Assert.AreEqual(i, c.CurrentValue);
//            }
//            sw.Stop();
//            Console.WriteLine("Elapsed msec: {0}", sw.ElapsedMilliseconds - 50);
//            Console.WriteLine("Ops: {0}", Math.Round(0.000001 * count * 1000.0 / (sw.ElapsedMilliseconds * 1.0), 2));

//        }


//        [Test]
//        public void CouldEnumerateChangingSM() {
//            var count = 1000000;
//            var sw = new Stopwatch();
//            sw.Start();
//            var sm = new SortedMap<DateTime, double>();
//            //sm.IsSynchronized = true;
//            var c = sm.GetCursor();

//            for (int i = 0; i < count; i++) {
//                sm.Add(DateTime.UtcNow.Date.AddSeconds(i), i);
//                var version = sm.Version;
//                //if (i > 10) {
//                //    sm.Add(DateTime.UtcNow.Date.AddSeconds(i - 10), i - 10 + 1);
//                //    Assert.IsTrue(sm.Version > version);
//                //}
//                c.MoveNext();
//                Assert.AreEqual(i, c.CurrentValue);
//            }
//            sw.Stop();
//            Console.WriteLine("Elapsed msec: {0}", sw.ElapsedMilliseconds - 50);
//            Console.WriteLine("Ops: {0}", Math.Round(0.000001 * count * 1000.0 / (sw.ElapsedMilliseconds * 1.0), 2));

//        }

//        [Test]
//        public void CouldMoveAtGE() {
//            var sm = new SortedMap<int, int>(50);
//            for (int i = 0; i < 100; i++) {
//                sm[i] = i;
//            }
//            Assert.IsTrue(sm.IsRegular);
//            Assert.AreEqual(99, sm.rkLast);
//            var cursor = sm.GetCursor();

//            cursor.MoveAt(-100, Lookup.GE);

//            Assert.AreEqual(0, cursor.CurrentKey);
//            Assert.AreEqual(0, cursor.CurrentValue);

//            var shouldBeFalse = cursor.MoveAt(-100, Lookup.LE);
//            Assert.IsFalse(shouldBeFalse);

//        }

//        [Test]
//        public void CouldMoveAtFromDict() {

//            var sm = new SortedMap<int, int>(new Dictionary<int, int>()
//                {
//                    { 1, 1},
//                    //{ 2, 2},
//                    { 3, 3},
//                    //{ 4, 4},
//                    { 5, 5}
//                });

//            Assert.IsTrue(sm.IsRegular);
//            Assert.AreEqual(5, sm.rkLast);
//            var cursor = sm.GetCursor();

//            cursor.MoveAt(-100, Lookup.GE);

//            Assert.AreEqual(1, cursor.CurrentKey);
//            Assert.AreEqual(1, cursor.CurrentValue);

//            var shouldBeFalse = cursor.MoveAt(-100, Lookup.LE);
//            Assert.IsFalse(shouldBeFalse);

//        }


//        [Test]
//        public void CouldMoveAtLE() {
//            var scm = new SortedMap<long, long>();
//            for (long i = int.MaxValue; i < int.MaxValue * 4L; i = i + int.MaxValue) {
//                scm[i] = i;
//            }

//            var cursor = scm.GetCursor();

//            var shouldBeFalse = cursor.MoveAt(0, Lookup.LE);
//            Assert.IsFalse(shouldBeFalse);

//        }

//        //[Test]
//        //public void CouldSerializeSMWithSingleElement() {
//        //    var sm = new SortedMap<long, long>();
//        //    sm.Add(1, 1);
//        //    //Assert.Fail("TODO");
//        //    //var sm2 = BinarySerializer.Deserialize<SortedMap<long, long>>(Serialization.Serializer.Serialize(sm));
//        //    //Assert.AreEqual(1, sm2.Count);
//        //    //Assert.AreEqual(1, sm2.First.Value);
//        //    //Assert.AreEqual(1, sm2.First.Key);

//        //}


//        [Test]
//        public void IsSyncedIsSetAutomaticallyForCursor() {
//            var sm = new SortedMap<long, long>();
//            sm.Add(1, 1);

//            Task.Run(() => {
//                var c = sm.GetCursor();
//                c.MoveNext();
//            }).Wait();
//            Assert.IsTrue(sm.IsSynchronized);
//        }

//        [Test]
//        public void IsSyncedIsSetAutomaticallyForEnumerator() {
//            var sm = new SortedMap<long, long>();
//            sm.Add(1, 1);

//            Task.Run(() => {
//                foreach (var l in sm) {
//                    Assert.AreEqual(1, l.Value);
//                }
//            }).Wait();
//            Assert.IsTrue(sm.IsSynchronized);
//        }


//        [Test]
//        public void AddExistingThrowsAndKeepsVersion() {
//            var sm = new SortedMap<long, long>();
//            sm.Add(1, 1);
//            Assert.AreEqual(1, sm.Version);
//            Assert.Throws<ArgumentException>(() => sm.Add(1, 1));
//            Assert.AreEqual(1, sm.Version);
//        }

//    }
//}
